src/theory/symmetry_detect.h

```diff
@@ -0,0 +1,146 @@
+   * d_var_to_subvar is the mapping from the variable to the substitution
+   * variable, and d_subvar_to_vars is the mapping from the substitution
+   * variable to a list of variables forming a region of a partition.
+   */
+  class Partition
+  {
+   public:
+    /** Term corresponding to the partition, e.g., x + y = 0 */
+    Node d_term;
+    /** Mapping between the variable and the substitute variable x -> w, y -> w,
+     * z -> w */
+    std::map<Node, Node> d_var_to_subvar;
+
+    /** Mapping between the substitute variable and variables w-> { x, y, z } */
+    std::map<Node, std::vector<Node> > d_subvar_to_vars;
+  };
+
+  /**
+   * We build the partition trie indexed by
+   * parts[0].var_to_subvar[v]....parts[n].var_to_subvar[v]. The leaves of a
+   * partition trie is the new regions of a partition
+   */
+  class PartitionTrie
+  {
+   public:
+    /** Variables at the leave */
+    std::vector<Node> d_variables;
+
+    /** The mapping from a node to its children */
+    std::map<Node, PartitionTrie> d_children;
+
+    /** Add variable v to the trie, indexed by
+     * parts[0].var_to_subvar[v]....parts[n].var_to_subvar[v]. */
+    void addNode(Node v, std::vector<Partition>& parts);
+    void addNode(Node v, std::vector<Node>& subs);
+
+    /** Get all the new regions of a partition and store in part */
+    void getNewPartition(Partition& part);
+    void getNewPartition(Partition& part, PartitionTrie& pt);
+  };
+
+
+  /** True and false constant nodes */
+  Node d_trueNode;
+  Node d_falseNode;
+
+  /** Cache for partitions */
+  std::map<Node, Partition> d_term_partition;
+
+  /** detect
+   *
+   * Called on the input assertions, where assertions are interpreted as a
+   * conjunction. This computes a partition P of the variables in assertions
+   * such that for each set S in P, then all automorphisms for S applied to
+   * assertions result in an equisatisfiable formula.
+   */
+  Partition detect(std::vector<Node>& assertions);
+
+  /** Find symmetries in node */
+  SymmetryDetect::Partition findPartitions(Node node);
+
+  /** Collect children of a node
+   *  If the kind of node is associative, we will chain its children together.
+   *  We might need optimizations here, such as rewriting the input to negation
+   * normal form.
+   * */
+  void collectChildren(Node node, std::vector<Node>& children);
+  void collectChildren(Node node, Kind k, std::vector<Node>& children);
+
+  /** Print a partition */
+  void printPartition(Partition p);
+
+  /** Get all variables from partitions */
+  void getVariables(std::vector<Partition>& partitions,
+                    std::unordered_set<Node, NodeHashFunction>& vars);
+
+  /** Process singleton partitions and add all variables to vars */
+  void processSingletonPartitions(std::vector<Partition>& partitions,
+                                  std::unordered_set<Node, NodeHashFunction>& vars);
+
+  /** Do matches on singleton partitions */
```

Comment:
Can you say more about what this function does?  This should give a clear idea of what the arguments are and how they are updated.